src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs

```diff
@@ -27,28 +27,89 @@ internal class SnappyCompressor : ICompressor
     {
         public CompressorType Type => CompressorType.Snappy;
 
+        /* Declared in header order as defined in the spec:
+         * https://github.com/mongodb/specifications/blob/7461d48df580b2e32b33fdc01188c86fc79773c1/source/compression/OP_COMPRESSED.rst#op_compressed*/
+        private const int MessageLengthSize = sizeof(int);
+        private const int RequestIdSize = sizeof(int);
+        private const int ResponseToSize = sizeof(int);
+        private const int OpCodeSize = sizeof(int);
+        private const int OriginalOpCodeSize = OpCodeSize;
+        private const int UncompressedSizeSize = sizeof(int);
+        private const int CompressorTypeSize = sizeof(byte); // compressorId in spec
+
+        /// <summary>
+        /// Compresses the message  in <paramref name="input"/>, writing the compressed message to
+        /// <paramref name="output"/>.
+        /// </summary>
+        /// <param name="input"> The input stream. Stream MUST HAVE read the exactly up to and including the
+        /// originalOpCode.</param>
+        /// <param name="output">The output stream. Stream MUST HAVE written the header exactly up to and including
+        /// compressorType aka compressorId.</param>
+        /// <remarks>
+        /// The precondition for <paramref name="input"/> mirrors the precondition for <see cref="Decompress"/>'s output
+        /// precondition.
+        /// The precondition for <paramref name="output"/> mirrors the precondition for <see cref="Decompress"/>'s input
+        /// precondition.
+        /// </remarks>
         public void Compress(Stream input, Stream output)
         {
 #if NET452 || NETSTANDARD2_0
-            using (var snappyStream = new SnappyStream(output, CompressionMode.Compress, true))
-            {
-                input.EfficientCopyTo(snappyStream);
-            }
+            // extract uncompressedSize from OP_COMPRESSED header in output
+            const int uncompressedSizeOffset = -(UncompressedSizeSize + CompressorTypeSize);
+            var position = output.Position;
+            output.Seek(uncompressedSizeOffset, SeekOrigin.Current); // rewind head
+            var uncompressedSize = ReadInt32(output);
+            output.Position = position; // return head to original position
+
+            var uncompressedMessage = new byte[uncompressedSize]; // does not include uncompressed message headers
+            input.ReadBytes(uncompressedMessage, offset: 0, count: uncompressedSize, CancellationToken.None);
+            var maxSize = SnappyCodec.GetMaxCompressedLength(uncompressedSize);
+            var compressedBytes = new byte[maxSize];
+            var compressedSize = SnappyCodec.Compress(uncompressedMessage, 0, uncompressedSize, compressedBytes, 0);
+            output.Write(compressedBytes, 0, compressedSize);
```

Comment:
I don't like the idea of *reading* from `output`.

Will this work instead:

        public void Compress(Stream input, Stream output)
        {
    #if NET452 || NETSTANDARD2_0
            var uncompressedSize = (int)(input.Length - input.Position);
            var uncompressedBytes = new byte[uncompressedSize]; // does not include uncompressed message headers
            input.ReadBytes(uncompressedBytes, offset: 0, count: uncompressedSize, CancellationToken.None);
            var maxCompressedSize = SnappyCodec.GetMaxCompressedLength(uncompressedSize);
            var compressedBytes = new byte[maxCompressedSize];
            var compressedSize = SnappyCodec.Compress(uncompressedBytes, 0, uncompressedSize, compressedBytes, 0);
            output.Write(compressedBytes, 0, compressedSize);
    #else
            throw new NotSupportedException();
    #endif
        }

Also... let's try and reserve word "message" for when talking about the entire message, including headers.

Reply:
Re: "message": ack. 
Your suggestion is definitely cleaner and works. Fixed!